File: =EXAMTIME/1_TradeDesk/City Subway Stations/c.py
costs = {
    'A': {'A': 'AB', 'B': 'AB', 'C': 'ABC'},
    'B': {'A': 'AB', 'B': 'AB', 'C': 'BC'},
    'C': {'A': 'ABC', 'B': 'BC', 'C': 'BC'}
}

start_city = 'Z'
dest_city = 'X'

city_names_A = ['X']
city_names_B = ['Z']
city_names_C = ['Y']

if start_city in city_names_A:
    start_city = 'A'
elif start_city in city_names_B:
    start_city = 'B'
elif start_city in city_names_C:
    start_city = 'C'

if dest_city in city_names_A:
    dest_city = 'A'
elif dest_city in city_names_B:
    dest_city = 'B'
elif dest_city in city_names_C:
    dest_city = 'C'

def calc_cost(costs, start, dest):
    if start not in costs or dest not in costs:
        return None
    return costs[start][dest]

cost = calc_cost(costs, start_city, dest_city)

if cost is not None:
    print(f"Cost from {start_city} to {dest_city}: {cost}")
else:
    print("Invalid input cities.")

File: =EXAMTIME/1_TradeDesk/City Subway Stations/c.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class c {

    // Function to calculate the cost based on the provided costs map
    public static String calculateCost(Map<String, Map<String, String>> costs, List<String> cityNames, String startCity, String destinationCity) {
        if (!cityNames.contains(startCity) || !cityNames.contains(destinationCity)) {
            // Return null if either the start or destination city is not in the cityNames list
            return null;
        }

        // Get the positions of start and destination cities in the cityNames list
        int startIdx = cityNames.indexOf(startCity);
        int destIdx = cityNames.indexOf(destinationCity);

        // Get the cost from the start city to the destination city
        String cost = costs.get(cityNames.get(startIdx)).get(cityNames.get(destIdx));

        return cost;
    }

    public static void main(String[] args) {
        // Example: List of city names (X, Y, Z) for each set of costs
        List<String> cityNamesA = Arrays.asList("X", "Y", "Z");
        List<String> cityNamesB = Arrays.asList("X", "Y", "Z");
        List<String> cityNamesC = Arrays.asList("X", "Y", "Z");

        // Example: List of city names (A, B, C) and their associated costs
        Map<String, Map<String, String>> costs = new HashMap<>();
        
        // Cost from A to A, B, C
        costs.put("A", Map.of("A", "AB", "B", "AB", "C", "ABC"));

        // Cost from B to A, B, C
        costs.put("B", Map.of("A", "AB", "B", "AB", "C", "BC"));

        // Cost from C to A, B, C
        costs.put("C", Map.of("A", "ABC", "B", "BC", "C", "BC"));

        // Example usage
        String startCity = "C";
        String destinationCity = "A";

        // Choose the appropriate cityNames list based on the cost set
        List<String> cityNames = cityNamesC;

        String cost = calculateCost(costs, cityNames, startCity, destinationCity);

        if (cost != null) {
            System.out.println("Cost from " + startCity + " to " + destinationCity + ": " + cost);
        } else {
            System.out.println("Invalid input cities.");
        }
    }
}
